python/tank/bootstrap/manager.py

```diff
@@ -221,11 +244,50 @@ def set_progress_callback(self, callback):
         """
         self._progress_cb = callback
 
-    def bootstrap_engine(self, engine_name, entity=None):
+    def bootstrap_engine(self,
+                         engine_name,
+                         entity=None,
+                         progress_callback=None,
+                         completed_callback=None,
+                         failed_callback=None):
         """
         Create an sgtk instance for the given project or site,
         then launch into the given engine.
 
+        When property ``async_bootstrap`` is false, the whole engine bootstrap logic
+        will be executed synchronously in the main application thread. When true,
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
+        When property ``async_bootstrap`` is true, 3 callback functions should be provided.
```

Comment:
As mentioned before, usage of a property will make it easy to add control of the bootstrap logic using a config file value, and makes it easy to fallback from asynchronous to synchronous when Qt is not available.

From my point of view, trying to combine the 3 callbacks into just one would make the callback logic less obvious to understand, and would complicate writing the client code considering how different the 3 callback signatures are.

Though, I have updated the synchronous engine bootstrap logic in method `bootstrap_engine` to also use the 3 callback functions when they are provided. This way, client code can run the engine bootstrap synchronously or asynchronously without code changes, just by setting the value of property `async_bootstrap`.